Create file values in API v2

benjamingeer opened the pull request:
As per https://github.com/dhlab-basel/Knora/issues/998#issuecomment-426695230:

- [ ] Add Sipi routes to process file uploads
- [ ] Implement file value creation in Knora

subotic commented:
@benjamingeer since you're working on the Sipi Lua scripts, could you maybe implement #874?

benjamingeer commented:
@subotic Yes, I'll try to do that, too.

subotic commented:
thanks :-)

subotic commented:
Sure, I will be tomorrow in the office and will coordinate with him.

tobiasschweizer commented:
Ok, great. See you tomorrow!

benjamingeer commented on webapi/src/main/scala/org/knora/webapi/OntologyConstants.scala:
```diff
@@ -816,7 +816,6 @@ object OntologyConstants {
         val LinkValueHasSourceIri: IRI = KnoraApiV2PrefixExpansion + "linkValueHasSourceIri"
 
         val FileValueAsUrl: IRI = KnoraApiV2PrefixExpansion + "fileValueAsUrl"
-        val FileValueIsPreview: IRI = KnoraApiV2PrefixExpansion + "fileValueIsPreview"
```
With IIIF, it doesn't make sense to have a separate preview image. If you want a preview, you just request the same image in a smaller size. The separate preview image in the old SALSAH was needed only because we didn't have IIIF yet.

benjamingeer commented on knora-ontologies/knora-base.ttl:
```diff
@@ -2390,7 +2390,7 @@
                                      [ rdf:type owl:Restriction ;
                                        owl:onProperty :qualityLevel ;
-                                       owl:cardinality "1"^^xsd:nonNegativeInteger
+                                       owl:maxCardinality "1"^^xsd:nonNegativeInteger
```
See "For a subsequent PR" in the description of this PR.

tobiasschweizer replied:
What if someone creates data using v2 in the meantime?

benjamingeer replied:
Then I think it won't work. If I'm not mistaken, I'm going to have to change v1 to fake the "preview" image if it doesn't exist. But this PR is already very large (110 files changed). Wouldn't you rather review that fix in a separate PR?

benjamingeer replied:
Anyway, we don't even have a beta release of v2 yet. Only daredevils like you are using it. :)

tobiasschweizer replied:
I think for v2, we just need one`hasStillImageFileValue` anyway (see 
https://github.com/dhlab-basel/Knora/pull/1011#discussion_r235290491).

My suggestion: change that back to  `owl:cardinality "1"` and return `fileValueIsPreview` for v2 for this PR. Then make a new Knora PR with a tracking Knora UI core PR. I think I have to adapt a few things regarding image handling in the viewer module, not just in the API format conversion.

tobiasschweizer replied:
> Anyway, we don't even have a beta release of v2 yet. Only daredevils like you are using it. :)

But we a have a productive system https://beol.dasch.swiss which I do not want to break when doing the next import.

benjamingeer replied:
> return fileValueIsPreview for v2 for this PR.

You mean for **all** images? How would that make sense? All images would be previews?

tobiasschweizer replied:
> You mean for all images? How would that make sense? All images would be previews?

just the current behaviour: if you get two images, you know which one is the preview. Then we can go on and just return one image as IIIF.

benjamingeer replied:
> if you get two images, you know which one is the preview

But you don't understand. Two images are not going to be uploaded. Only one.

tobiasschweizer replied:
maybe it is easier if you call me on facetime.

benjamingeer replied:
Just talked to @lrosenth: he says that there is no need to generate a separate preview image for efficiency: Sipi can efficiently return the preview on the fly in response to a IIIF request, and Sipi also has an internal cache.

tobiasschweizer replied:
Ok. Please document that until Api v1 is adapted, `qualityLevel` will be generated by v1 imports whereas it won't with v2 imports. During this time, the triplestore's content is not version independent regarding the property `qualityLevel`.

Existing data also needs to be converted once v1 is changed. Preferably there will be a conversion routine that handles this.

benjamingeer replied:
#1068

subotic replied:
I think that we should be rather safe than sorry and disable the new routes in `V2` that don't generate `qualityLevel` until #1068 is resolved.

benjamingeer replied:
> disable the new routes

OK. Should I just have these routes throw `NotImplementedException`?

benjamingeer replied:
But then how would I test them in an e2e or integration test?

benjamingeer commented:
After more discussion with @subotic: I'll make sure that the full-size image file values created by v2 are the same (in the triplestore) as the ones created by v1, so the data is consistent at least in that respect.

benjamingeer commented:
There is also `knora-base:valueHasQname`, which appears a few times in the test data (not consistently). API v1 writes it to the triplestore with every file value. It seems to be redundant with `isPreview`, and Knora never reads it from the triplestore.

benjamingeer commented:
And `valueHasQname` doesn't even exist in `knora-base`!

tobiasschweizer commented:
Once API v2 and v1 behave the same,

```
rdfs:subClassOf :Representation ,
                                [ rdf:type owl:Restriction ;
                                  owl:onProperty :hasStillImageFileValue ;
                                  owl:minCardinality "1"^^xsd:nonNegativeInteger
                                ] ;
```
this would have to be changed to cardinality 1, right?

Actually, we could never correctly express or check that we expect two file values.

benjamingeer commented:
> this would have to be changed to cardinality 1, right?

I believe that a `Representation` is supposed to be something that can have multiple file values attached to it. If I remember correctly, lukas's idea was that you could have a black-and-white image, a colour image, an X-ray image, all representing the same thing in different ways, and all part of the same `Representation`. I think this is why the `Representation` class exists.

benjamingeer commented:
There wasn't even a `valueHasQname` in knora-base when we first moved the Knora code to GitHub.

benjamingeer commented:
> If I remember correctly, Lukas's idea was that you could have a black-and-white image, a colour image, an X-ray image, all representing the same thing in different ways

But maybe we need to rethink this, because the problem would be that there would be no way to separately annotate or link to these different images. You could only annotate the `Representation`. So do we really need the `Representation` class at all?

benjamingeer commented:
There isn't even a way to put extra metadata in each file value to say who the photographer was for each image, or anything like that.

benjamingeer commented:
So I guess we need `Representation` because it's the only place where you could put extra metadata. For example, `incunabula:page` is a subclass of `StillImageRepresentation`, and `page` has the metadata about the image. But it doesn't actually make sense for a representation to have more than one file value, because then there would be no way to give metadata for the individual files. So yes, I agree with you, we should change those cardinalities to `owl:cardinality 1`.

benjamingeer commented:
@tobiasschweizer @subotic I changed API v1 to generate IIIF preview URLs, ignoring the preview image file values in the triplestore. I updated the test data so the tests pass. I browsed around in the SALSAH 1 GUI and it looks like it works (IIIF preview images are requested and displayed), but I haven't been able to get the SALSAH 1 GUI tests to work, because of timeouts and other errors that I don't understand.

tobiasschweizer commented:
@loicjaouen had the same problem (
https://github.com/dhlab-basel/Knora/pull/1018#issuecomment-434718882).

My guess is that there were ontology changes that break the tests. In the tests, everything works with positions (n-th property).

If you are here next week, we could have a look at it.

benjamingeer commented:
> If you are here next week, we could have a look at it.

Yes, thanks, I'll be there.

tobiasschweizer commented:
Ok, great. Mostly it is really trivial. You just have to adapt the positions. But it is very cumbersome since the GUI was written in a testing-hostile manner ;-)

subotic commented:
The BEOL Import Test run without problems. 2000 letters with images where imported.

subotic commented on webapi/src/main/scala/org/knora/webapi/responders/v1/ResourcesResponderV1.scala:
```diff
@@ -943,23 +943,19 @@ class ResourcesResponderV1 extends Responder {
+                            // a IIIF preview URL from the full-size image.
 
-                            val locations: Option[Seq[LocationV1]] = sourceObj.fileValues.find(fileVal => fileVal.permissioCode.nonEmpty && !fileVal.image.isPreview) match {
+                            val (preview: Option[LocationV1], locations: Option[Seq[LocationV1]]): (Option[LocationV1], Option[Seq[LocationV1]]) = sourceObj.fileValues.find(fileVal => fileVal.permissionCode.nonEmpty && !fileVal.image.isPreview) match {
```
Will this still work if there is no preview image in the triplestore, so the SPARQL query only returns one row per resource?

benjamingeer replied:
Not sure, will check.

benjamingeer replied:
Yes, it works: I deleted two preview images from the incunabula test data, and the context test still passes.

subotic commented on webapi/src/main/scala/org/knora/webapi/responders/v2/SipiResponderV2.scala:
```diff
@@ -0,0 +1,184 @@
+/*
```
In a future PR, this could be moved out of responders.

subotic commented on webapi/src/test/scala/org/knora/webapi/responders/v2/ValuesResponderV2Spec.scala:
```diff
@@ -1742,18 +1746,168 @@ class ValuesResponderV2Spec extends CoreSpec() with ImplicitSender {
+                dimY = 256
+            )
+
+            // Knora will accept this request.
```
...but the mock sipi responder will fail.

benjamingeer replied:
https://github.com/dhlab-basel/Knora/pull/1011/commits/f588bd05e55e6b4b3efa4eb91eaed718acd65c22

benjamingeer commented on sipi/scripts/jwt.lua:
```diff
@@ -0,0 +1,80 @@
+        return nil
+    end
+
+    -- TODO: test with a token signed with the wrong secret.
```
Yes: https://github.com/dhlab-basel/Knora/pull/1011/files#diff-48d76fbef5e1c09276c1a4da1e7eb2b0R217

benjamingeer commented on sipi/scripts/upload.lua:
```diff
@@ -0,0 +1,157 @@
+for image_index, image_params in pairs(server.uploads) do
+    -- Create the temporary directory if necessary.
+
+    local temp_dir = config.imgroot .. '/tmp/'
```
https://github.com/dhlab-basel/Knora/pull/1011/commits/f588bd05e55e6b4b3efa4eb91eaed718acd65c22

benjamingeer commented on sipi/scripts/jwt.lua:
```diff
@@ -0,0 +1,80 @@
+function get_knora_token()
+    local token = get_token()
+
+    if token == nil then
```
https://github.com/dhlab-basel/Knora/pull/1011/commits/f588bd05e55e6b4b3efa4eb91eaed718acd65c22

benjamingeer commented:
@subotic I think this is OK to merge now.

subotic commented:
This pull request has been mentioned on **Discuss DaSCH**. There might be relevant details there:

https://discuss.dasch.swiss/t/support-non-image-files-in-knora-api-v2/33/1
